File: bae-android/app/src/main/java/fm/bae/app/BaeCrashReporting.kt
package fm.bae.app

import android.content.Context
import io.sentry.Sentry
import io.sentry.android.core.SentryAndroid

private const val CRASH_REPORTING_TAG = "bae.CrashReporting"
private val crashReportingLogger = BaeLogger(CRASH_REPORTING_TAG)

object BaeCrashReporting {
    fun configure(context: Context) {
        // The baeium edition ships without crash reporting by design, so its
        // absence is expected and silent.
        if (BuildConfig.BAE_EDITION != "bae") return

        val dsn = configuredBuildString(BuildConfig.BAE_SENTRY_DSN)
        val environment = configuredBuildString(BuildConfig.BAE_ENVIRONMENT)
        val gitCommit = configuredBuildString(BuildConfig.BAE_GIT_COMMIT)
        if (dsn == null || environment == null || gitCommit == null) {
            crashReportingLogger.warning(
                "Crash reporting disabled: Sentry not configured at build time " +
                    "(need BAE_SENTRY_DSN, BAE_ENVIRONMENT, BAE_GIT_COMMIT). " +
                    "Crashes will not be reported.",
            )
            return
        }

        try {
            SentryAndroid.init(context) { options ->
                options.dsn = dsn
                options.environment = environment
                options.release = "bae@${BuildConfig.VERSION_NAME}+$gitCommit"
                options.dist = gitCommit
                options.isSendDefaultPii = false
                options.isEnableNdk = true
            }
            Sentry.configureScope { scope ->
                scope.setTag("edition", BuildConfig.BAE_EDITION)
                scope.setTag("git_commit", gitCommit)
            }
        } catch (e: Exception) {
            crashReportingLogger.error("Failed to configure crash reporting", e)
        }
    }
}

File: BaeKit/Sources/BaeKit/BaeCrashReporting.swift
import Foundation
import Sentry

public enum BaeCrashReporting {
    public static func configure(edition: AppEdition) {
        guard let config = config(edition: edition) else { return }
        SentrySDK.start { options in
            options.dsn = config.dsn
            options.environment = config.environment
            options.releaseName = config.releaseName
            options.dist = config.gitCommit
            options.sendDefaultPii = false
        }
        SentrySDK.configureScope { scope in
            scope.setTag(value: config.edition, key: "edition")
            scope.setTag(value: config.gitCommit, key: "git_commit")
        }
    }

    private static func config(edition: AppEdition) -> CrashReportingConfig? {
        guard edition == .bae,
            let dsn = BuildInfo.infoString("BaeSentryDsn"),
            let environment = BuildInfo.infoString("BaeEnvironment"),
            let appVersion = BuildInfo.infoString("CFBundleShortVersionString"),
            let gitCommit = BuildInfo.infoString("BaeGitCommit")
        else {
            return nil
        }
        return CrashReportingConfig(
            dsn: dsn,
            environment: environment,
            releaseName: "bae@\(appVersion)+\(gitCommit)",
            edition: edition.rawValue,
            gitCommit: gitCommit
        )
    }
}

private struct CrashReportingConfig {
    let dsn: String
    let environment: String
    let releaseName: String
    let edition: String
    let gitCommit: String
}
